File: benchmarks/72-dogleg/dogleg.m
(* Experiment 72 -- dogleg (Powell dogleg trust region) unconstrained
   optimization (FindMinimum, Method -> "Dogleg"; alias "dogleg").

   MEASURES src/numerical_calculus/findmin.c's dogleg -- it builds the dense
   Hessian and blends the Cauchy (steepest-descent) point with the Newton point
   along the dogleg path, intersected with the trust radius -- against
   scipy.optimize.minimize(method="dogleg"). scipy is given the analytic Jacobian
   AND Hessian, so the race is solver-vs-solver (both dogleg trust region).

   dogleg needs a POSITIVE-DEFINITE model Hessian (scipy raises LinAlgError
   otherwise), so unlike experiments 71/73/74/75 the problem set is PD
   THROUGHOUT: ill-conditioned diagonal quadratics (condition 1, 1e2, 1e4, 1e6,
   constant SPD Hessian) and the Trid function (constant tridiagonal SPD Hessian,
   integer optimum). The check is the objective at the optimum rounded to 10^6:
   0 for the f*=0 quadratics, the exact integer optimum for Trid (n=10 -> -210,
   n=20 -> -1520). Objectives are built with Sum/Total over fresh global symbols
   and passed through Evaluate[] past FindMinimum's HoldAll. *)

Get["../harness.m"];

require[{"FindMinimum"}];


quad[n_, c_] := With[{v = Table[Symbol["qv" <> ToString[i]], {i, 1, n}]},
                     Sum[10^(c (i - 1)/(n - 1)) v[[i]]^2, {i, 1, n}]];
quadS[n_]  := Table[{Symbol["qv" <> ToString[i]], 1.0}, {i, 1, n}];
trid[n_]   := With[{v = Table[Symbol["tv" <> ToString[i]], {i, 1, n}]},
                   Sum[(v[[i]] - 1)^2, {i, 1, n}] - Sum[v[[i]] v[[i - 1]], {i, 2, n}]];
tridS[n_]  := Table[{Symbol["tv" <> ToString[i]], 0.0}, {i, 1, n}];

(* Precompute out of the timed region. *)
q0  = quad[10, 0]; q0s  = quadS[10];
q2  = quad[10, 2]; q2s  = quadS[10];
q4  = quad[10, 4]; q4s  = quadS[10];
q6  = quad[10, 6]; q6s  = quadS[10];
t10 = trid[10];    t10s = tridS[10];
t20 = trid[20];    t20s = tridS[20];

bench["01 Sphere n10", FindMinimum[Evaluate[q0], Evaluate[q0s], Method -> "Dogleg", MaxIterations -> 5000];];
check["01 Sphere n10", Round[10^6 First[FindMinimum[Evaluate[q0], Evaluate[q0s], Method -> "Dogleg", MaxIterations -> 5000]]]];

bench["02 Illcond quad n10 c1e2", FindMinimum[Evaluate[q2], Evaluate[q2s], Method -> "Dogleg", MaxIterations -> 5000];];
check["02 Illcond quad n10 c1e2", Round[10^6 First[FindMinimum[Evaluate[q2], Evaluate[q2s], Method -> "Dogleg", MaxIterations -> 5000]]]];

bench["03 Illcond quad n10 c1e4", FindMinimum[Evaluate[q4], Evaluate[q4s], Method -> "Dogleg", MaxIterations -> 5000];];
check["03 Illcond quad n10 c1e4", Round[10^6 First[FindMinimum[Evaluate[q4], Evaluate[q4s], Method -> "Dogleg", MaxIterations -> 5000]]]];

bench["04 Illcond quad n10 c1e6", FindMinimum[Evaluate[q6], Evaluate[q6s], Method -> "Dogleg", MaxIterations -> 5000];];
check["04 Illcond quad n10 c1e6", Round[10^6 First[FindMinimum[Evaluate[q6], Evaluate[q6s], Method -> "Dogleg", MaxIterations -> 5000]]]];

bench["05 Trid n10", FindMinimum[Evaluate[t10], Evaluate[t10s], Method -> "Dogleg", MaxIterations -> 5000];];
check["05 Trid n10", Round[10^6 First[FindMinimum[Evaluate[t10], Evaluate[t10s], Method -> "Dogleg", MaxIterations -> 5000]]]];

bench["06 Trid n20", FindMinimum[Evaluate[t20], Evaluate[t20s], Method -> "Dogleg", MaxIterations -> 5000];];
check["06 Trid n20", Round[10^6 First[FindMinimum[Evaluate[t20], Evaluate[t20s], Method -> "Dogleg", MaxIterations -> 5000]]]];

File: benchmarks/72-dogleg/dogleg.py
#!/usr/bin/env python3
"""Experiment 72 -- dogleg unconstrained optimization (scipy column).

Same 6 problems as ``dogleg.m``, same labels, same order, solved with
scipy.optimize.minimize(method="dogleg") and the analytic Jacobian AND Hessian,
so the race is solver-vs-solver (both Powell dogleg trust region).

dogleg needs a POSITIVE-DEFINITE model Hessian (scipy raises LinAlgError
otherwise), so the problem set is PD throughout: ill-conditioned diagonal
quadratics (condition 1, 1e2, 1e4, 1e6) and the Trid function (constant SPD
Hessian). The check is the objective at the optimum rounded to 10^6 (0 for the
f*=0 quadratics, the exact integer for Trid).
"""

import sys, os; sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

import numpy as np
from scipy.optimize import minimize

from harness import bench, check, require

require(["scipy.optimize:minimize"])

_METHOD = "dogleg"
_OPTS = {"maxiter": 5000, "gtol": 1e-10}


def _solve(fun, jac, hess, x0):
    return minimize(fun, np.asarray(x0, float), jac=jac, hess=hess,
                    method=_METHOD, options=_OPTS)


# ---- ill-conditioned diagonal quadratic  f = sum 10^(c(i-1)/(n-1)) x_i^2 ----
def _quad_case(n, c):
    a = np.array([10.0 ** (c * i / (n - 1)) for i in range(n)])
    fun = lambda v: float(np.dot(a, v * v))
    jac = lambda v: 2.0 * a * v
    hess = lambda v: np.diag(2.0 * a)
    x0 = np.ones(n)
    return lambda: _solve(fun, jac, hess, x0)


# ---- Trid: f = sum (x_i-1)^2 - sum x_i x_{i-1}; min -n(n+4)(n-1)/6 ----
def _trid_case(n):
    def fun(v):
        return float(np.sum((v - 1.0) ** 2) - np.sum(v[1:] * v[:-1]))
    def jac(v):
        g = 2.0 * (v - 1.0)
        g[:-1] -= v[1:]
        g[1:] -= v[:-1]
        return g
    def hess(v):
        H = 2.0 * np.eye(n)
        for i in range(n - 1):
            H[i, i + 1] = -1.0
            H[i + 1, i] = -1.0
        return H
    x0 = np.zeros(n)
    return lambda: _solve(fun, jac, hess, x0)


_q0, _q2, _q4, _q6 = _quad_case(10, 0), _quad_case(10, 2), _quad_case(10, 4), _quad_case(10, 6)
_t10, _t20 = _trid_case(10), _trid_case(20)

bench("01 Sphere n10", _q0)
check("01 Sphere n10", int(np.floor(1e6 * _q0().fun + 0.5)))

bench("02 Illcond quad n10 c1e2", _q2)
check("02 Illcond quad n10 c1e2", int(np.floor(1e6 * _q2().fun + 0.5)))

bench("03 Illcond quad n10 c1e4", _q4)
check("03 Illcond quad n10 c1e4", int(np.floor(1e6 * _q4().fun + 0.5)))

bench("04 Illcond quad n10 c1e6", _q6)
check("04 Illcond quad n10 c1e6", int(np.floor(1e6 * _q6().fun + 0.5)))

bench("05 Trid n10", _t10)
check("05 Trid n10", int(np.floor(1e6 * _t10().fun + 0.5)))

bench("06 Trid n20", _t20)
check("06 Trid n20", int(np.floor(1e6 * _t20().fun + 0.5)))
